overlord/patch/patch4.go

```diff
@@ -0,0 +1,171 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patch
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+func init() {
+	patches[4] = patch4
+}
+
+type patch4T struct{} // for namespacing of the helpers
+
+func (patch4T) snapSetupAndState(task *state.Task) (*snapstate.SnapSetup, *snapstate.SnapState, error) {
```

Comment:
notice that using snapstate stuff in patches comes with a bit of perils, see patch2 for example introduces its own OldSnapSetup and OldSnapState, maybe things are settled enough but the trade off here is having to rewrite this if we change something significant there that breaks code in the patch